File: JS/twoNumSum/twoNumSum3.js
// O(n log(n)) time | O(1) space
const twoNumSum = (array, targetSum) => {

  // sort the given array so we can find the pair
  // working from the outer edges
  array.sort();
  // above defaults to: array.sort((a, b) => a - b);
  let lowIndex = 0;
  let highIndex = array.length - 1;
	
	// iterate through the input array
	while (lowIndex < highIndex) {
    const currentSum = array[lowIndex] + array[highIndex];
    
    if (currentSum === targetSum) {
      return [array[lowIndex], array[highIndex]];
    } else if (currentSum < targetSum) {
      lowIndex++;
    } else if (currentSum > targetSum) {
      highIndex--;
    }
  }
  // return empty array if no pair is found
  return [];
}

// Testing
let array = [5, 2, 7, 4, 1, 6, 3, 8, 9];
let targetSum = 17;

console.log(twoNumSum(array, targetSum));

File: Python/two_num_sum/two_num_sum3.py
# O(n) time | O(1) space
def two_num_sum(array, targetSum):
  # sort the input array to search faster & save space
  array.sort()
  left_idx = 0
  right_idx = len(array) - 1
  
  while left_idx < right_idx:
    current_total = array[left_idx] + array[right_idx]
    
    if current_total == targetSum:
      return [array[left_idx], array[right_idx]]
    elif current_total < targetSum:
      left_idx += 1
    else:
      right_idx -= 1
  
  # return an empty array if no pair is found
  return []


# Testing
ex = [3, 5, -4, 8, 11, 1, -1, 6]
print(two_num_sum(ex, 10))
